### PR TITLE
Reorderable list

### PR DESCRIPTION
**Pull request by tbelaire**
This has some buttons which can move around, but they keep the same Id.

It would have been nice if I could borrow self.data inside the callback immutably, as I first tried it:
https://github.com/tbelaire/crochet/blob/wishful-thinking/examples/travelling_button.rs#L30

I can see that `self.data` is already borrowed immutably, so I don't see why I cant borrow it immutably again.  It'd be nice to avoid this kinda ugly pattern if possible.

Anyways, this moving of rows around is something fairly reasonable for the gui to do.  It suffers from the same bug in #5, as swapping the data doesn't seem to trigger an update.

I know you were aiming for the Optimistic Merge pattern, but I don't actually trust my code to be idiomatic, and would rather have some code review if you don't mind.

### CODE REVIEW

**Comment by raphlinus** [comment]
Some thoughts.

First, it's a known limitation of the current tree mutation engine that it doesn't do reorders, only insert, delete, and update. I do plan to add this, as it is useful functionality. What happens on swap is that the first element is deleted, the second element is skipped, then another copy of the first element is inserted. That's fine for labels but otherwise you'd lose the widget state. As I say, fixable, but it does mean making the tree mutation protocol more complicated. It's kinda nice that there are 4 variants: insert, delete, update, skip. I have a design in mind with "detach" and "reattach" as 2 more variants, but am still refining it in my head.

You can fix the borrow pattern by doing `let data = &self.data;` before calling the list's run method, then referring to `data` rather than `self.data`. The reason it's complaining here is that the closure wants to borrow all of `self`, and there's a mutable borrow from `self` for its `list_view` field. I think this is standard Rust borrow checker being difficult and nothing to do with Crochet. Even though the use of closures for scopes in general is still a bit controversial (see #1), for a list view component, there kinda needs to be a closure for creating a list element. That said, the pattern you've come up with - not doing a lot of work inside the closure, but getting an "action" out of it, seems like a pretty general solution.

I'd be happy to merge what I have, but I'll let you decide whether to revise your PR based on this feedback.

**Comment by tbelaire** [comment]
Yeah I think I'd like to simplify it down to only one action being passed out.

**Comment by tbelaire** [comment]
Alright, fixed it up, thanks for the review.

**Review by raphlinus — APPROVED** [review]
Looks good, thanks! I'm adding you to the committer list so you can merge future PR's.

**Pull request merged**
